commit a4cd9b05c2a9d78d3e78581a13d5ed890bc85efb (refs/changes/27/2727/4)
Author: Pedro Palazon Candel <pedro@joyent.com>
Date:   2017-10-09T16:36:12+02:00 (2 years ago)
    
    PAPI-143 Moray buckets need to be reindexed after schema updates
    Reviewed by: Marsell Kukuljevic <marsell@joyent.com>

diff --git a/lib/backend.js b/lib/backend.js
index 2f68328..c9b1ff4 100644
--- a/lib/backend.js
+++ b/lib/backend.js
@@ -15,13 +15,13 @@
 
 var moray  = require('moray');
 var assert = require('assert');
-var util   = require('util');
 var jsprim = require('jsprim');
 var VError = require('verror');
 
 var RETRY_BUCKET_CREATION_INTERVAL = 10000; // in ms
 var DB_RESERVED_NAMES = ['default', 'group'];
-
+var MAX_ROWS_TO_REINDEX = 100;
+var MAX_REINDEX_ATTEMPTS = 5;
 
 /*
  * Backend constructor. Returns a set of methods useful for fetching and
@@ -60,6 +60,40 @@ function (cb) {
     var backend = createMorayClient(self.opts);
     self.backend = backend;
 
+    var reindexAttempts = 0;
+
+    function reindexObjects(done) {
+        backend.reindexObjects(bucket, MAX_ROWS_TO_REINDEX,
+            function reindexCb(reindexErr, reindexRes) {
+            reindexAttempts += 1;
+            if (reindexErr) {
+                log.error({
+                    err: reindexErr,
+                    bucket: bucket
+                }, 'Reindex bucket error');
+                if (reindexAttempts < MAX_REINDEX_ATTEMPTS) {
+                    reindexObjects(done);
+                } else {
+                    done(reindexErr);
+                }
+                return;
+            }
+
+            if (reindexRes.processed === 0) {
+                log.info('Reindex bucket done');
+                done();
+                return;
+            }
+
+            log.info({
+                bucket: bucket,
+                results: reindexRes
+            }, 'reindexed %d rows', reindexRes.processed);
+
+            reindexObjects(done);
+        });
+    }
+
     /*
      * Ensure bucket for packages in Moray exists. Keep retrying every 10s
      * until successful.
@@ -97,7 +131,7 @@ function (cb) {
                     morayConnectCallback();
                 }, RETRY_BUCKET_CREATION_INTERVAL);
             } else {
-                cb();
+                reindexObjects(cb);
             }
         });
     };
